components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTx.java

```diff
@@ -208,16 +209,52 @@ long countComponents(@Nullable String whereClause,
   void deleteBucket(Bucket bucket);
 
   /**
-   * Creates a new Blob.
+   * Creates a new Blob and updates the given asset with a reference to it, hash metadata, size, and content type.
+   * The old blob, if any, will be deleted.
+   *
+   * @param asset               the {@link Asset} that should reference newly created blob.
+   * @param blobName            blob name (must not be unique), but it will be used also in content validation. See
+   *                            {@link ContentValidator}.
+   * @param inputStream         the content to be streamed into blob store.
+   * @param hashAlgorithms      {@link HashAlgorithm}s to be applied while streaming to blob store, will be set
+   *                            in {@link Asset} attributes.
+   * @param headers             optional, custom headers for blob, if any.
+   * @param declaredContentType optional, the declared MIME type of the blob. See {@link ContentValidator}.
+   * @return {@link BlobRef} to the newly created blob. As side effect, passed in {@link Asset} is modified too, but is
+   * unsaved. Caller must ensure {@link #saveAsset(Asset)} is invoked before this TX ends.
    */
-  BlobRef createBlob(InputStream inputStream, Map<String, String> headers);
+  BlobRef setBlob(Asset asset,
```

Comment:
This method is basically `createBlob` plus `attachBlob` with a minor precondition, so I wonder if we really need it compared to just calling those methods directly. Also it should ideally return `AssetBlob` for API consistency with `attachBlob`.